New rule: Content triggered on hover is hoverable (#ep1s13) (#1396)
act-rules/act-rules.github.io: pages/glossary/content-that-becomes-visible.md

---
title: Content that becomes visible
key: content-that-becomes-visible
unambiguous: true
objective: true
input_aspects:
  - CSS styling
  - DOM tree
---

The _content that becomes visible_ after an [event][] [firing][] is any element for which all of the following is true:

> - 500 milliseconds after the [event][] [firing][] the element has [visible text content][]; and

Reviewer: We need to capture an assumption that content on hover will appear within 500ms.

> - a tick before the [event][] [firing][] there was no element with the same [descendant text content][] that the element has after the [event][] [firing][]; and

Reviewer: Seems to me like this would apply to elements that update their content too. If I have a button that changes its text on hover, it 1. has visible content after the event. and 2. does not have the same descendant text content. Additionally, descendant text content does not take shadow DOM into account.

---
title: Content that becomes visible
key: content-that-becomes-visible
unambiguous: true
objective: true
input_aspects:
  - CSS styling
  - DOM tree
---

> The _content that becomes visible_ is the [root][] element, if one exists, of the [tree][] that contains all elements that, as result of an [event][] [firing][], meet any of the following:

Reviewer: I'm not too sure what is going on here. How can a tree not have a root element? Do you maybe mean to find the closest common ancestor? How do I know that something is the "result of an event firing"? I don't think we should try to figure out if something is or isn't the result of an event. I think it's sufficient to determine that the content appears within half a second of firing the event, and that the content part of a region that updates without user input.

Author: The "if one exists" was meant to apply to the tree, not the root element. My bad.

And I'll update to use the half a second instead of resulting from. That's definitely more testable!

> - the element did not exist on the [web page][] prior to the [event][] [firing][] and has [visible text content][]; or

Reviewer: I think there is a fundamental problem in this. If I have an event rerenders the page, all elements are new, even if nothing changed. `body.innerHTML = body.innerHTML` will do that. Modern frameworks try to avoid unnecessary rerenders, but there is a lot of code on the web that rerenders components, irrespective of if something actually changed.

I think we need to define this in a way that works with rerenders. I took a stab at doing that just for text nodes, here's what I've come up with. That isn't enough, but I hope it gets helps move this forward.

```md
A text node is considered to be _visible after an event_ if both of the following of the following are true:

- A tick before the event, there was no [visible][] text node with the same node value; and
- 500 milliseconds after the [event][] [firing][], there the text node is [visible][].
```

Author: I don't think that works because it requires that there are no text nodes with the same node value. If the new content includes a text node with the same value of an already present text node, then it will miss the new content.
I'll try to think of a solution for this...

Author: Following your suggestion, check if you think that what I came up with works

> - the element has [visible text content][] and is added to the [web page][] where the [event][] was [fired][firing]; or
> - the element [attributes][]' [values][] are modified in a way that cause some or all of its [text nodes][] to become [visible][].

Reviewer: Visible text content isn't bound to the closest ancestor. Any text nodes added to the page results in the HTML element having visible text content added.

How about something like:

```suggestion
has visible text content that did not exist on the page prior to the event firing.
```

Author: We don't need to worry about the bounding because this tree only contains new elements (therefore elements that were there already are not considered) and elements that have had their attributes modified (therefore if their text nodes changed without a change in their attributes, they are not considered)
